File: sdk/wire.ts
import {
  createHash,
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  randomBytes,
  sign as edSign,
  verify as edVerify,
} from 'node:crypto';
import canonicalizeImport from 'canonicalize';
import { parseTree, type Node, type ParseError } from 'jsonc-parser';
import { z } from 'zod';

export const PROFILE = 'passport.v0.ed25519-jcs';
export const PROGRAM = 'passport.hidden-limit.bp5.ristretto.32x2.v1';
const canonicalize = canonicalizeImport as unknown as (input: unknown) => string | undefined;
export const id = z.string().regex(/^[a-zA-Z0-9][a-zA-Z0-9:._-]{0,127}$/);
export const timestamp = z.number().int().positive().max(Number.MAX_SAFE_INTEGER);
export const amount = z.number().int().positive().max(4294967295);
export const scopeSchema = z.object({
  actions: z.array(id).min(1).max(32),
  resources: z.array(id).min(1).max(32),
  audiences: z.array(id).min(1).max(32),
});
export const requestSchema = z
  .object({
    request_id: id,
    issuer_org: id,
    agent_id: id,
    runtime_key_id: id,
    delegation_id: id,
    action: id,
    resource: id,
    audience: id,
    recipient: id,
    amount_minor_units: amount,
    currency: z.literal('USD'),
    asset: z.literal('iso4217:USD'),
    network: z.literal('simulation'),
    method: z.literal('POST'),
    path: z.literal('/v0/purchases'),
    body: z.record(z.unknown()),
    nonce: id.min(32),
    issued_at: timestamp,
    expires_at: timestamp,
    want_proof: z.boolean(),
  })
  .strict();
export type RequestPayload = z.infer<typeof requestSchema>;
export interface Doc<T = Record<string, unknown>> {
  profile: typeof PROFILE;
  kind: string;
  key_id: string;
  payload: T;
  signature: string;
}
export interface KeyPair {
  private_key: string;
  public_key: string;
}
export interface RuntimeKey {
  id: string;
  org_id: string;
  agent_id: string;
  public_key: string;
  delegation_id: string;
  not_before: number;
  expires_at: number;
}
export interface Delegation extends z.infer<typeof scopeSchema> {
  id: string;
  org_id: string;
  agent_id: string;
  runtime_key_id: string;
  not_before: number;
  expires_at: number;
}
export interface PolicyKey extends z.infer<typeof scopeSchema> {
  id: string;
  org_id: string;
  public_key: string;
  max_ttl: number;
  not_before: number;
  expires_at: number;
}
export interface Policy {
  id: string;
  org_id: string;
  version: number;
  predecessor: string;
  commitment: string;
  predicate: string;
  currency: string;
  asset: string;
  network: string;
  not_before: number;
  expires_at: number;
}
export interface Capability {
  capability_id: string;
  issuer_org: string;
  agent_id: string;
  runtime_key_id: string;
  delegation_id: string;
  action: string;
  resource: string;
  audience: string;
  recipient: string;
  amount_minor_units: number;
  currency: string;
  asset: string;
  network: string;
  request_digest: string;
  nonce: string;
  issued_at: number;
  expires_at: number;
  policy_version: string;
  policy_commitment: string;
  proof_program: string;
}
export interface Proof {
  program: string;
  proof: string;
}
export interface Evidence {
  request: Doc<RequestPayload>;
  delegation: Doc<Delegation>;
  capability: Doc<Capability>;
  proof?: Proof;
}
export interface Pins {
  registry_key_id: string;
  registry_public_key: string;
  trusted_roots: Record<string, Record<string, string>>;
}
export interface TrustBundle {
  org_id: string;
  challenge: string;
  issued_at: number;
  expires_at: number;
  active_policy: string;
  roots: Array<{ key_id: string; public_key: string; revoked: boolean }>;
  records: Array<{ document: Doc<any>; revoked: boolean }>;
}
export interface Verification {
  decision: 'allow' | 'deny';
  code: string;
  request_digest: string;
  evidence_digest: string;
  policy_version: string;
  state_issued_at: number;
  valid_until: number;
  trace: string[];
}
export interface Receipt {
  id: string;
  receipt_type: 'authorization' | 'execution' | 'denial';
  org_id: string;
  decision: string;
  code: string;
  request_id: string;
  capability_id: string;
  verifier: string;
  issued_at: number;
  policy_version: string;
  request_digest: string;
  evidence_digest: string;
  trace: string[];
}
export const envelopeSchema = z
  .object({
    profile: z.literal(PROFILE),
    kind: id,
    key_id: id,
    payload: z.record(z.unknown()),
    signature: z.string().regex(/^[A-Za-z0-9_-]{86}$/),
  })
  .strict();
export const evidenceSchema = z
  .object({
    request: envelopeSchema,
    delegation: envelopeSchema,
    capability: envelopeSchema,
    proof: z
      .object({
        program: z.literal(PROGRAM),
        proof: z
          .string()
          .regex(/^[a-f0-9]+$/)
          .max(8192),
      })
      .strict()
      .optional(),
  })
  .strict();

export class PassportError extends Error {
  constructor(
    public code: string,
    public status = 403,
    public detail?: unknown,
  ) {
    super(code);
  }
}
export function assert(condition: unknown, code: string): asserts condition {
  if (!condition) throw new PassportError(code);
}
export function canonical(value: unknown): string {
  const text = canonicalize(value);
  if (text === undefined) throw new PassportError('invalid_request', 400);
  return text;
}
export function digest(value: unknown): string {
  return createHash('sha256').update(canonical(value)).digest('hex');
}
export function now(): number {
  return Math.floor(Date.now() / 1000);
}
export function randomID(prefix: string): string {
  return `${prefix}-${randomBytes(16).toString('hex')}`;
}
export function keys(): KeyPair {
  const pair = generateKeyPairSync('ed25519');
  return {
    private_key: pair.privateKey.export({ format: 'pem', type: 'pkcs8' }).toString(),
    public_key: pair.publicKey.export({ format: 'der', type: 'spki' }).toString('base64url'),
  };
}
export function sign<T>(kind: string, key_id: string, payload: T, privateKey: string): Doc<T> {
  const header = { profile: PROFILE, kind, key_id, payload };
  return {
    ...header,
    profile: PROFILE,
    signature: edSign(null, Buffer.from(canonical(header)), createPrivateKey(privateKey)).toString(
      'base64url',
    ),
  };
}
export function verify<T>(document: Doc<T>, kind: string, publicKey: string): T {
  assert(envelopeSchema.safeParse(document).success && document.kind === kind, 'invalid_signature');
  const { signature, ...header } = document;
  try {
    assert(
      Buffer.from(signature, 'base64url').toString('base64url') === signature &&
        Buffer.from(publicKey, 'base64url').toString('base64url') === publicKey,
      'invalid_signature',
    );
    const key = createPublicKey({
      key: Buffer.from(publicKey, 'base64url'),
      format: 'der',
      type: 'spki',
    });
    assert(
      key.asymmetricKeyType === 'ed25519' &&
        edVerify(null, Buffer.from(canonical(header)), key, Buffer.from(signature, 'base64url')),
      'invalid_signature',
    );
  } catch {
    throw new PassportError('invalid_signature');
  }
  return document.payload;
}
export function strictJSON(text: string): unknown {
  const errors: ParseError[] = [];
  const tree = parseTree(text, errors, { allowTrailingComma: false, disallowComments: true });
  if (!tree || errors.length) throw new PassportError('invalid_request', 400);
  function visit(n: Node): void {
    if (n.type === 'object') {
      const names = new Set<string>();
      for (const p of n.children ?? []) {
        const key = p.children?.[0]?.value;
        if (names.has(key)) throw new PassportError('invalid_request', 400);
        names.add(key);
      }
    }
    for (const child of n.children ?? []) visit(child);
  }
  visit(tree);
  return JSON.parse(text);
}
export function inScope(scope: z.infer<typeof scopeSchema>, r: RequestPayload): boolean {
  return (
    scope.actions.includes(r.action) &&
    scope.resources.includes(r.resource) &&
    scope.audiences.includes(r.audience)
  );
}
export function validWindow(from: number, to: number): boolean {
  return (
    Number.isSafeInteger(from) &&
    Number.isSafeInteger(to) &&
    from > 0 &&
    from <= now() &&
    now() < to
  );
}

export async function http<T>(url: string, body?: unknown, token?: string): Promise<T> {
  const response = await fetch(url, {
    method: body === undefined ? 'GET' : 'POST',
    headers: {
      ...(body !== undefined ? { 'Content-Type': 'application/json' } : {}),
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: body === undefined ? undefined : canonical(body),
    signal: AbortSignal.timeout(8000),
  }).catch(() => {
    throw new PassportError('unavailable', 503);
  });
  const text = await response.text();
  let data: any;
  try {
    data = strictJSON(text);
  } catch {
    throw new PassportError('unavailable', 503);
  }
  if (!response.ok) throw new PassportError(data.code ?? 'denied', response.status, data);
  return data as T;
}
export async function fetchTrust(registry: string, org: string, pins: Pins): Promise<TrustBundle> {
  assert(id.safeParse(org).success && pins.trusted_roots[org], 'untrusted_issuer');
  const challenge = randomBytes(32).toString('hex');
  const doc = await http<Doc<TrustBundle>>(
    `${registry}/v0/organizations/${encodeURIComponent(org)}/trust?challenge=${challenge}`,
  );
  assert(doc.key_id === pins.registry_key_id, 'invalid_signature');
  const bundle = verify(doc, 'trust_bundle', pins.registry_public_key);
  assert(
    bundle.org_id === org &&
      bundle.challenge === challenge &&
      bundle.expires_at - bundle.issued_at <= 5 &&
      validWindow(bundle.issued_at, bundle.expires_at),
    'stale_state',
  );
  return bundle;
}
export function trustedRecord<T>(
  bundle: TrustBundle,
  pins: Pins,
  kind: string,
  id: string,
): Doc<T> {
  const entry = bundle.records.find(
    (e) => e.document.kind === kind && e.document.payload.id === id,
  );
  assert(entry, 'denied');
  assert(!entry.revoked, 'revoked');
  const keyID = entry.document.key_id,
    publicKey = pins.trusted_roots[bundle.org_id]?.[keyID];
  const root = bundle.roots.find((r) => r.key_id === keyID);
  assert(root && !root.revoked, 'revoked');
  assert(publicKey && root.public_key === publicKey, 'untrusted_issuer');
  verify(entry.document, kind, publicKey);
  return entry.document as Doc<T>;
}
export function verifyRuntime(
  request: Doc<RequestPayload>,
  delegation: Doc<Delegation>,
  bundle: TrustBundle,
  pins: Pins,
): RuntimeKey {
  const r = requestSchema.parse(request.payload);
  assert(r.issuer_org === bundle.org_id, 'untrusted_issuer');
  trustedRecord(bundle, pins, 'organization', r.issuer_org);
  const agent = trustedRecord<{ id: string; org_id: string }>(
    bundle,
    pins,
    'agent',
    r.agent_id,
  ).payload;
  const key = trustedRecord<RuntimeKey>(bundle, pins, 'runtime_key', r.runtime_key_id).payload;
  const d = trustedRecord<Delegation>(bundle, pins, 'delegation', r.delegation_id);
  assert(
    agent.org_id === r.issuer_org &&
      key.org_id === r.issuer_org &&
      key.agent_id === r.agent_id &&
      key.delegation_id === r.delegation_id,
    'denied',
  );
  assert(
    d.payload.org_id === r.issuer_org &&
      d.payload.agent_id === r.agent_id &&
      d.payload.runtime_key_id === r.runtime_key_id &&
      digest(d) === digest(delegation),
    'denied',
  );
  assert(request.key_id === key.id, 'invalid_signature');
  verify(request, 'request', key.public_key);
  assert(
    validWindow(key.not_before, key.expires_at) &&
      validWindow(d.payload.not_before, d.payload.expires_at) &&
      validWindow(r.issued_at, r.expires_at),
    'expired',
  );
  assert(
    r.issued_at >= key.not_before &&
      r.issued_at >= d.payload.not_before &&
      r.expires_at - r.issued_at <= 120 &&
      r.expires_at <= key.expires_at &&
      r.expires_at <= d.payload.expires_at &&
      inScope(d.payload, r),
    'denied',
  );
  return key;
}

File: pkg/wire/wire.go
// Package wire defines Passport's versioned, interoperable signing profile.
package wire

import (
	"bytes"
	"crypto/ed25519"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"regexp"

	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
)

const Profile = "passport.v0.ed25519-jcs"
const ProofProgram = "passport.hidden-limit.bp5.ristretto.32x2.v1"
const MaxAmount = uint64(4294967295)

type Document struct {
	Profile   string          `json:"profile"`
	Kind      string          `json:"kind"`
	KeyID     string          `json:"key_id"`
	Payload   json.RawMessage `json:"payload"`
	Signature string          `json:"signature"`
}
type Header struct {
	Profile string          `json:"profile"`
	Kind    string          `json:"kind"`
	KeyID   string          `json:"key_id"`
	Payload json.RawMessage `json:"payload"`
}
type Organization struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	RootKeyID string `json:"root_key_id"`
	PublicKey string `json:"public_key"`
	CreatedAt int64  `json:"created_at"`
}
type Agent struct {
	ID        string `json:"id"`
	OrgID     string `json:"org_id"`
	Name      string `json:"name"`
	CreatedAt int64  `json:"created_at"`
}
type RuntimeKey struct {
	ID           string `json:"id"`
	OrgID        string `json:"org_id"`
	AgentID      string `json:"agent_id"`
	PublicKey    string `json:"public_key"`
	DelegationID string `json:"delegation_id"`
	NotBefore    int64  `json:"not_before"`
	ExpiresAt    int64  `json:"expires_at"`
}
type Scope struct {
	Actions   []string `json:"actions"`
	Resources []string `json:"resources"`
	Audiences []string `json:"audiences"`
}
type Delegation struct {
	ID           string `json:"id"`
	OrgID        string `json:"org_id"`
	AgentID      string `json:"agent_id"`
	RuntimeKeyID string `json:"runtime_key_id"`
	Scope
	NotBefore int64 `json:"not_before"`
	ExpiresAt int64 `json:"expires_at"`
}
type PolicyKey struct {
	ID        string `json:"id"`
	OrgID     string `json:"org_id"`
	PublicKey string `json:"public_key"`
	Scope
	MaxTTL    int64 `json:"max_ttl"`
	NotBefore int64 `json:"not_before"`
	ExpiresAt int64 `json:"expires_at"`
}
type Policy struct {
	ID          string `json:"id"`
	OrgID       string `json:"org_id"`
	Version     int64  `json:"version"`
	Predecessor string `json:"predecessor"`
	Commitment  string `json:"commitment"`
	Predicate   string `json:"predicate"`
	Currency    string `json:"currency"`
	Asset       string `json:"asset"`
	Network     string `json:"network"`
	NotBefore   int64  `json:"not_before"`
	ExpiresAt   int64  `json:"expires_at"`
}
type Request struct {
	RequestID    string          `json:"request_id"`
	IssuerOrg    string          `json:"issuer_org"`
	AgentID      string          `json:"agent_id"`
	RuntimeKeyID string          `json:"runtime_key_id"`
	DelegationID string          `json:"delegation_id"`
	Action       string          `json:"action"`
	Resource     string          `json:"resource"`
	Audience     string          `json:"audience"`
	Recipient    string          `json:"recipient"`
	Amount       uint64          `json:"amount_minor_units"`
	Currency     string          `json:"currency"`
	Asset        string          `json:"asset"`
	Network      string          `json:"network"`
	Method       string          `json:"method"`
	Path         string          `json:"path"`
	Body         json.RawMessage `json:"body"`
	Nonce        string          `json:"nonce"`
	IssuedAt     int64           `json:"issued_at"`
	ExpiresAt    int64           `json:"expires_at"`
	WantProof    bool            `json:"want_proof"`
}
type Capability struct {
	CapabilityID     string `json:"capability_id"`
	IssuerOrg        string `json:"issuer_org"`
	AgentID          string `json:"agent_id"`
	RuntimeKeyID     string `json:"runtime_key_id"`
	DelegationID     string `json:"delegation_id"`
	Action           string `json:"action"`
	Resource         string `json:"resource"`
	Audience         string `json:"audience"`
	Recipient        string `json:"recipient"`
	Amount           uint64 `json:"amount_minor_units"`
	Currency         string `json:"currency"`
	Asset            string `json:"asset"`
	Network          string `json:"network"`
	RequestDigest    string `json:"request_digest"`
	Nonce            string `json:"nonce"`
	IssuedAt         int64  `json:"issued_at"`
	ExpiresAt        int64  `json:"expires_at"`
	PolicyVersion    string `json:"policy_version"`
	PolicyCommitment string `json:"policy_commitment"`
	ProofProgram     string `json:"proof_program"`
}
type Proof struct {
	Program string `json:"program"`
	Proof   string `json:"proof"`
}
type Evidence struct {
	Request    Document `json:"request"`
	Delegation Document `json:"delegation"`
	Capability Document `json:"capability"`
	Proof      *Proof   `json:"proof,omitempty"`
}
type Root struct {
	KeyID     string `json:"key_id"`
	PublicKey string `json:"public_key"`
	Revoked   bool   `json:"revoked"`
}
type Record struct {
	Document Document `json:"document"`
	Revoked  bool     `json:"revoked"`
}
type Bundle struct {
	OrgID        string   `json:"org_id"`
	Challenge    string   `json:"challenge"`
	IssuedAt     int64    `json:"issued_at"`
	ExpiresAt    int64    `json:"expires_at"`
	ActivePolicy string   `json:"active_policy"`
	Roots        []Root   `json:"roots"`
	Records      []Record `json:"records"`
}
type Revocation struct {
	ID         string `json:"id"`
	OrgID      string `json:"org_id"`
	TargetKind string `json:"target_kind"`
	TargetID   string `json:"target_id"`
	IssuedAt   int64  `json:"issued_at"`
}
type Rotation struct {
	ID            string `json:"id"`
	OrgID         string `json:"org_id"`
	PreviousKeyID string `json:"previous_key_id"`
	PublicKey     string `json:"public_key"`
	IssuedAt      int64  `json:"issued_at"`
}
type Receipt struct {
	ID             string   `json:"id"`
	ReceiptType    string   `json:"receipt_type"`
	OrgID          string   `json:"org_id"`
	Decision       string   `json:"decision"`
	Code           string   `json:"code"`
	RequestID      string   `json:"request_id"`
	CapabilityID   string   `json:"capability_id"`
	Verifier       string   `json:"verifier"`
	IssuedAt       int64    `json:"issued_at"`
	PolicyVersion  string   `json:"policy_version"`
	RequestDigest  string   `json:"request_digest"`
	EvidenceDigest string   `json:"evidence_digest"`
	Trace          []string `json:"trace"`
}

var identifier = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9:._-]{0,127}$`)

func ID(s string) bool { return identifier.MatchString(s) }
func Canonical(v any) ([]byte, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	return jsoncanonicalizer.Transform(raw)
}

// Decode rejects duplicate keys, unknown struct fields, and trailing data.
func Decode(raw []byte, v any) error {
	if _, err := jsoncanonicalizer.Transform(raw); err != nil {
		return err
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.DisallowUnknownFields()
	if err := dec.Decode(v); err != nil {
		return err
	}
	if err := dec.Decode(new(any)); err != io.EOF {
		return errors.New("trailing JSON")
	}
	return nil
}
func Payload[T any](d Document) (T, error) { var p T; err := Decode(d.Payload, &p); return p, err }
func Digest(v any) string {
	b, err := Canonical(v)
	if err != nil {
		return ""
	}
	h := sha256.Sum256(b)
	return hex.EncodeToString(h[:])
}
func PublicKey(encoded string) (ed25519.PublicKey, error) {
	der, err := base64.RawURLEncoding.Strict().DecodeString(encoded)
	if err != nil {
		return nil, err
	}
	v, err := x509.ParsePKIXPublicKey(der)
	if err != nil {
		return nil, err
	}
	key, ok := v.(ed25519.PublicKey)
	if !ok {
		return nil, errors.New("only Ed25519 is supported")
	}
	return key, nil
}
func Sign(kind, keyID string, payload any, key ed25519.PrivateKey) (Document, error) {
	p, err := Canonical(payload)
	if err != nil {
		return Document{}, err
	}
	h := Header{Profile, kind, keyID, p}
	b, err := Canonical(h)
	if err != nil {
		return Document{}, err
	}
	return Document{Profile, kind, keyID, p, base64.RawURLEncoding.EncodeToString(ed25519.Sign(key, b))}, nil
}
func Verify(d Document, kind, public string) error {
	if d.Profile != Profile || d.Kind != kind || !ID(d.KeyID) {
		return errors.New("invalid signing profile or kind")
	}
	key, err := PublicKey(public)
	if err != nil {
		return err
	}
	sig, err := base64.RawURLEncoding.Strict().DecodeString(d.Signature)
	if err != nil {
		return err
	}
	b, err := Canonical(Header{d.Profile, d.Kind, d.KeyID, d.Payload})
	if err != nil {
		return err
	}
	if !ed25519.Verify(key, b, sig) {
		return errors.New("invalid signature")
	}
	return nil
}
func InScope(s Scope, action, resource, audience string) bool {
	contains := func(xs []string, s string) bool {
		for _, x := range xs {
			if x == s {
				return true
			}
		}
		return false
	}
	return contains(s.Actions, action) && contains(s.Resources, resource) && contains(s.Audiences, audience)
}
func Window(now, from, to int64) bool { return from > 0 && from <= now && now < to }
func ValidRequest(r Request, now int64) error {
	if !ID(r.RequestID) || !ID(r.IssuerOrg) || !ID(r.AgentID) || !ID(r.RuntimeKeyID) || !ID(r.DelegationID) || !ID(r.Nonce) {
		return errors.New("invalid identifier")
	}
	if r.Amount == 0 || r.Amount > MaxAmount || r.Currency != "USD" || r.Asset != "iso4217:USD" || r.Network != "simulation" {
		return errors.New("invalid amount or units")
	}
	if len(r.Nonce) < 32 || r.ExpiresAt-r.IssuedAt > 120 || !Window(now, r.IssuedAt, r.ExpiresAt) {
		return fmt.Errorf("invalid request window")
	}
	if len(r.Body) == 0 || bytes.Equal(r.Body, []byte("null")) {
		return errors.New("missing body")
	}
	return nil
}
